proposals/0000-unify-type-level-literals.rst

```diff
@@ -0,0 +1,156 @@
+
+Proposed Change Specification
+-----------------------------
+The following is relative to the current behavior of ``DataKinds``.
+
+Grammar changes
+^^^^^^^^^^^^^^^
+The lexical syntax would gain the following new case:
+
+::
+
+    signedint -> ['+'|'-'] integer
+
+The ``atype`` non-terminal would gain new cases as follows:
+
+::
+
+    atype -> ...
+           | string
+           | char
+           | integer
+           | signedint
+           | float
+
+This is relative to the specification, not the current effects of ``DataKinds``.
+
+Semantics
+^^^^^^^^^
+The kinds of type-level literals will be as follows:
+
+::
+
+    "foo" :: String
+    'f' :: Char
+    -123 :: Integer
+    +123 :: Integer
+    123 :: Natural
+    3.14 :: Rational
+
+While the ``Rational`` literals are the most dubious, they require very minimal changes (since ``Ratio Integer`` will now work correctly "for free") and are needed to satisfy the original impetus for this change.
+
+The case of positive ``Integer`` literals is challenging. The syntax above is not great, but the obvious alternatives are a subkinding relation or an explicit conversion and I'd consider both of those to be worse.
+
+To handle the legacy case, ``Symbol`` and ``Nat`` will become aliases for ``String`` and ``Natural``, respectively.
+
+All of the mathematical type families in ``GHC.TypeNats`` will be generalized to be poly-kinded so that they work for these new numeric kinds, with the exception of a ``Rational`` implementation for  ``Div``, ``Mod``, and ``Log2``. ``/`` and ``%`` will be added for ``Rational``. Put more explicitly, their interfaces will be as listed in the block below, where ``a`` means any of ``Integer``, ``Natural``, or ``Rational`` and ``b`` means either ``Integer`` or ``Rational``.
+
+::
+
+    type family Add a (m :: a) (n :: a) :: a
+    type family Mul a (m :: a) (n :: a) :: a
+    type family Exp a b (m :: a) (n :: b) :: a
+    type family Sub a (m :: a) (n :: a) :: a
+    type family DivI a (m :: a) (n :: a) :: a
+    type family ModI a (m :: a) (n :: a) :: a
+    type family Log2I b (m :: b) :: b
+
+    type (m :: a) + (n :: a) = Add a m n
+    type (m :: a) * (n :: a) = Mul a m n
+    type (m :: a) ^ (n :: b) = Exp a b m n
+    type (m :: a) - (n :: a) = Sub a m n
+    type (m :: a) / (n :: a) = DivI a m n
+    type (m :: a) % (n :: a) = ModI a m n
+    type Div (m :: b) (n :: b) = DivI b m n
+    type Mod (m :: b) (n :: b) = ModI b m n
+    type Log2 (m :: b) = Log2I b m
+
+A new type family will be added, ``Data.Type.Equality.Cmp``, using a similar alias-wrapper trick to provide a uniform interface for comparisons.
+
+::
+
+    type family CmpI a (m :: a) (n :: a) :: Ordering
```

Comment:
Why do you need this type family? Making Cmp polymorphic is sufficient. Then the instances could look like:

```
type instance Cmp a b = CmpNat a b
type instance Cmp a b = CmpChar a b
type instance Cmp a b = CmpList a b
```

Type inference will do the rest.